packages/data-context/src/sources/ProjectDataSource.ts

```diff
@@ -1,10 +1,75 @@
-import type { CodeGenType, SpecType } from '@packages/graphql/src/gen/nxs.gen'
-import { FrontendFramework, FRONTEND_FRAMEWORKS, ResolvedFromConfig, RESOLVED_FROM, SpecFileWithExtension, STORYBOOK_GLOB } from '@packages/types'
+import type { CodeGenType } from '@packages/graphql/src/gen/nxs.gen'
+import { FoundSpec, FrontendFramework, FRONTEND_FRAMEWORKS, ResolvedFromConfig, RESOLVED_FROM, SpecFileWithExtension, STORYBOOK_GLOB } from '@packages/types'
 import { scanFSForAvailableDependency } from 'create-cypress-tests'
 import path from 'path'
+import Debug from 'debug'
+import commonPathPrefix from 'common-path-prefix'
+
+const debug = Debug('cypress:data-context')
 
 import type { DataContext } from '..'
-import type { Maybe } from '../data/coreDataShape'
+
+type SpecWithRelativeRoot = FoundSpec & { relativeToCommonRoot: string }
+
+interface MatchedSpecs {
+  projectRoot: string
+  testingType: Cypress.TestingType
+  specAbsolutePaths: string[]
+  specPattern: string | string[]
+}
+export function matchedSpecs ({
```

Comment:
This is the main "logical" change. 

Right now in the current production version of Cypress, if you have:

- `cypress/integration/foo.spec.js`
- `cypress/integration/foo/bar.spec.js`

The videos goes to:
- `cypress/videos/foo.spec.js.mp4`
- `cypress/videos/foo/bar.spec.js.mp4`

Screenshots:
- `cypress/screenshots/foo.spec.js/<screenshot.png>`
- `cypress/screenshots/foo/bar.spec.js/<screenshot.png>`

We always truncate the `cypress/integration` part of the `spec` when saving the artifact. Note it is `cypress/videos/foo.spec.js.mp4`, not `cypress/videos/cypress/integration/foo.spec.js.mp4`. This is easy using the `integrationFolder` property and just truncating it. We can't do this same thing with just a `specPattern`.

To maintain this behavior, ths code here finds the "common" path over all the specs and truncates it. Previously this was always `cypress/integration`. Now it's a bit harder to find.

There is some unsolvable edge cases. See this test: https://github.com/cypress-io/cypress/pull/19319/files#diff-59bf8023d3f96e2d91bc479be5372ebde0317cf6040e5e514ceb276ac71c1342R74-R87 OR https://github.com/cypress-io/cypress/blob/develop/system-tests/test/screenshot_nested_file_spec.js#L7.

Given `cypress/integration/nested-1/nested-2/foo.spec.js, there's no way to replicate the existing behavior - in the current system we'd just remove `cypress/integration`, which is the `integrationFolder`. This is no longer something we can do, since there's no such concept of a "integration folder" - just a glob pattern.

Reply:
Will this get weird as teams add tests over time?  Something like this:

*First Release*

File locations
* cypress/integration/a/b/c.spec.js
Screenshot locations
* cypress/screenshots/c.spec.js/<screenshot.png>

*Second Release*

File locations
* cypress/integration/a/b/c.spec.js
* cypress/integration/a/d/e.spec.js
Screenshot locations
* cypress/screenshots/b/c.spec.js/<screenshot.png>
* cypress/screenshots/d/e.spec.js/<screenshot.png>

So in the first release, screenshots will be in one location, but in a subsequent release they'll be in a different location (even though the file location hasn't changed).

Do we foresee that causing any issues?